Improved the current SDA workflow to reach the North American runs with 6400 sites.

DongchenZ opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->
<!--- Please select appropriate Priority, Status,and Type labels-->
<!--- If you do not have permission to select labels please state which labels you would like -->

## Description
<!--- Describe your changes in detail -->

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
1. The `foreach` package seems to be better compared to the `furrr` package concerning memory allocation. Thus, in this PR, I replaced every `furrr` with `foreach` during the general SDA workflow.
2. The computational power and memory are limited when executing certain SDA procedures locally (e.g., splitting meteorology files, writing configuration files, reading SDA outputs, running Bayesian MCMC analysis part, and removing files (e.g., removing NC files after the first SDA run)). Therefore, in the PR, I developed features of `qsub` submissions (specified by the `batch.settings` section in the XML file, which is also documented in the pecan book) during the SDA workflow.
3. To avoid the complex `if-else` usage within the current `sda.enkf_Multisite` function for the above `batch` job submissions, I developed a new `sda.enkf_NorthAmerica` function (although most of them are just copy-and-paste from the `sda.enkf_Multisite` function), which is cleaner and will only be used if the `batch.settings` is not empty.
<!--- If it fixes an open issue, please link to the issue here. -->

## Review Time Estimate
<!---When do you want your code reviewed by?-->
- [ ] Immediately
- [ ] Within one week
- [x] When possible

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue) <!-- please add issue number -->
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] My name is in the list of CITATION.cff
- [ ] I have updated the CHANGELOG.md.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

<!--this template is from https://www.talater.com/open-source-templates/#/page/99-->

Commit:
add the template for the MCMC qsub job.

Commit:
added namespaces

Commit:
added qsub parallel analysis functions.

Commit:
Merge branch 'develop' of https://github.com/DongchenZ/pecan into develop

Commit:
include qsub analysis part into the main function.

Commit:
Update namespace.

Commit:
Update dependencies.

Commit:
Update dependencies.

Commit:
Update namespace.

infotroph commented on modules/assim.sequential/NAMESPACE:
```diff
@@ -61,6 +63,7 @@ import(furrr)
 import(lubridate)
 import(nimble)
 importFrom(dplyr,"%>%")
+importFrom(foreach,"%dopar%")
```
Current CI failure translates to "You're not allowed to importFrom if the package is in Suggests rather than Imports"

DongchenZ replied:
It doesn't allow me to include more than 20 dependencies.

infotroph replied:
Three mutually exclusive ideas:
* Find an existing package in Imports that's easier to move to Suggests to make room for `foreach`
* Add the "more than 20 imports" message to tests/Rcheck_reference.log so CI checks ignore it
* Use `%dopar%` directly from the namespace rather than importFrom, which is hideously ugly but appears to work in this test: 
```
x = 1:5
val <- NULL
foreach::foreach(val = x) |> foreach::`%dopar%`(log(val))
```

infotroph commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -632,4 +639,99 @@ block.2.vector <- function (block.list, X, H) {
+    out <- system2(cmd, file.path(folder.path, "job.sh"), stdout = TRUE, stderr = TRUE)
+  }
+  # checking results.
+  PEcAn.logger::logger.info("Checking results.")
+  while (length(list.files(outdir, pattern="results.Rdata", recursive = T)) < folder.num) {
+    Sys.sleep(60)
+  }
```
Flagging that this has high potential to become an infinite loop if any job fails / is rejected by the queue / etc

infotroph replied:
More generally, I'm cautious of submitting anything to qsub and waiting for it to complete before returning from the submitting function -- Part of the point of qsub is that it does the waiting for you.

DongchenZ replied:
I am not sure if I resolve this correctly. Here, instead, I detect the job completion by checking if the job still exists on the server. This will ensure a finite `while` loop even if, for some reason, some jobs never get finished and are killed by the server.

infotroph replied:
> detect the job completion by checking if the job still exists on the server

👍 That seems like a solid improvement. -- still has the potential to be waiting a long time if the queue is stalled / moving slowly, but will only be waiting for outputs the server still plans to provide.

infotroph commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -632,4 +639,99 @@ block.2.vector <- function (block.list, X, H) {
+  PEcAn.logger::logger.info("Assembling results.")
+  analysis <- c()
+  for (path in seq_along(folder.paths)) {
+    load(file.path(path, "results.Rdata"))
+    analysis <- c(analysis, results)
```
```suggestion
    res_env <- new.env()
    load(file.path(path, "results.Rdata"), envir = res_env)
    analysis <- c(analysis, res_env$results)
```

DongchenZ replied:
Fixed.

infotroph commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -632,4 +639,99 @@ block.2.vector <- function (block.list, X, H) {
+  load(file.path(folder.path, "block.Rdata"))
+  # initialize parallel.
+  cores <- parallel::detectCores()
+  if (cores > 28) cores <- 28
```
This looks like a BU-specific number and should be configurable

DongchenZ replied:
Fixed.

infotroph commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -632,4 +639,99 @@ block.2.vector <- function (block.list, X, H) {
+  progress <- function(n) utils::setTxtProgressBar(pb, n)
+  opts <- list(progress=progress)
+  # parallel computation.
+  l <- NULL # fix GitHub check issue.
+  results <- NULL # fix GitHub check issue.
```
```suggestion
  l <- NULL # suppress R CMD check complaint (it doesn't know foreach assigns `l` below)
```

infotroph replied:
Note that I'm fairly sure the `results <- NULL` is'nt needed at all -- I bet the "no binding" complaint was from line 706 above, and my suggestion there should hopefully fix it.

DongchenZ replied:
Fixed.

Commit:
Apply suggestions.

infotroph commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -632,4 +641,117 @@ block.2.vector <- function (block.list, X, H) {
+  opts <- list(progress=progress)
+  # parallel computation.
+  l <- NULL # fix GitHub check issue.
+  results <- foreach::foreach(l = blocks, .packages=c("Kendall", "purrr"), .options.snow=opts) %dopar% {
+    MCMC_block_function(l)
+  }
```
Naive question: How closely equivalent is this to `parallel::parLapply(cl, blocks, MCMC_block_function)`?  Would it be worth considering that approach since it only uses existing dependencies?

DongchenZ replied:
So, I haven't used this parallel function before. The reason that I used `foreach` is that it has great management of memory usage with large data inputs. But I can try this function out to see if it performs similarly to the `foreach` package.

mdietze commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -632,4 +641,117 @@ block.2.vector <- function (block.list, X, H) {
+    dir.create(folder.path)
+    # save corresponding block list to the folder.
+    blocks <- block.list[head.num:tail.num]
+    save(blocks, file = file.path(folder.path, "block.Rdata"))
```
I find it hard to tell what the structure of `blocks` actually is, but is this something that could be saved as a csv dataframe or something similar to avoid Rdata? At the minimum I'd recommend switching to RDS, which is more transparent about what objects are being reloaded.

mdietze commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -632,4 +641,117 @@ block.2.vector <- function (block.list, X, H) {
+    blocks <- block.list[head.num:tail.num]
+    save(blocks, file = file.path(folder.path, "block.Rdata"))
+    # create job file.
+    jobsh <- readLines(con = system.file("analysis_qsub.job", package = "PEcAn.ModelName"), n = -1, warn=FALSE)
```
Why is the generic template model being hard-coded here? Do we really want to create a package dependency of the SDA module on the generic template model?

mdietze commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -632,4 +641,117 @@ block.2.vector <- function (block.list, X, H) {
+##' @title qsub_analysis_submission
+##' @param block.list list: MCMC configuration lists for the block SDA analysis.
+##' @param outdir character: SDA output path.
+##' @param job.per.folder numeric: number of jobs per folder.
```
This logic seems backwards to me. Generally with a HPC or cloud queueing system you want to specify the number of cores and/or nodes that you have, and then divide jobs across them.

DongchenZ replied:
based on folder numbers.

mdietze commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -632,4 +641,117 @@ block.2.vector <- function (block.list, X, H) {
+  for (i in 1:folder.num) {
+    # create folder for each set of job runs.
+    # calculate start and end index for the current folder.
+    head.num <- (i-1)*job.per.folder + 1
```
Are all jobs expected to take the same amount of time? If not (e.g. could one block have 1 site while another block has 1000 sites?), can we estimate which are expected to be longer or shorter so that we can load balance a bit more intelligently than doing so uniformly?

DongchenZ replied:
Should be balanced by number of sites within each block.

mdietze commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -632,4 +641,117 @@ block.2.vector <- function (block.list, X, H) {
+    jobsh <- gsub("@FOLDER_PATH@", folder.path, jobsh)
+    writeLines(jobsh, con = file.path(folder.path, "job.sh"))
+    # qsub command.
+    qsub <- "qsub -l h_rt=48:00:00 -l buyin -pe omp 28 -V -N @NAME@ -o @STDOUT@ -e @STDERR@ -S /bin/bash"
```
qsub settings are system and user specific, and thus need to be read from the `settings` object, not hard-coded. Also, I'd STRONGLY recommend setting up a single array-style qsub over submitting multiple jobs in a loop.

robkooper replied:
See [remote execution](https://pecanproject.github.io/pecan-documentation/master/xml-core-config.html#xml-host). Also a lot of this code seems to be in the PEcAn::remote package for example `start_qsub`

Commit:
Update dependency.

Commit:
Update dependency.

robkooper commented on models/template/inst/analysis_qsub.job:
```diff
@@ -0,0 +1,5 @@
+#!/bin/bash -l
+module load R/4.1.2
```
this is very server specific, in the pecan.xml we have a section for qsub that allows you to specify the modules that need to be loaded. See https://pecanproject.github.io/pecan-documentation/master/xml-core-config.html#xml-host for an example.

robkooper commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -632,4 +641,116 @@ block.2.vector <- function (block.list, X, H) {
+  # loop over sub-folders.
+  folder.paths <- job.ids <- c()
+  PEcAn.logger::logger.info(paste("Submitting", folder.num, "jobs."))
+  for (i in 1:folder.num) {
```
some systems will penalize you if you do many submissions in parallel, or only run two and then wait for those to be done, and more jobs you submit, the lower your priority in the queue. To overcome some of this I added the `modellauncher` this will create a text file with as first line the command to execute, and will have a list of folders it needs to run the command in.

Commit:
Pull from main branch.

Merge branch 'develop' of https://github.com/PecanProject/pecan into develop

# Conflicts:
#	modules/assim.sequential/DESCRIPTION

Commit:
Removing the job file template.

Commit:
automatic add-up.

Commit:
Update the qsub option to submit analysis jobs.

Commit:
Using foreach to create Y and R.

Commit:
Resolve dimension issue.

Commit:
Add functions for initializing qsub job submissions during the general SDA workflow.

Commit:
Create a new workflow solely for the North American SDA runs.

Commit:
Add qsub functions for parallel SDA workflows.

Commit:
Upgrade the weight calculation function and fix some bugs.

Commit:
Update the script for matching the current progress with NA SDA runs.

Commit:
Rd file.

Commit:
Add Rd files.

Commit:
Remove hard-coded path.

Commit:
Adding SDA workflow for NA SDA runs.

Commit:
Update documentation associated with the SDA batch job submission features.

Commit:
Update doc.

Commit:
Update dependency.

Commit:
Update dependency.

Commit:
Update job completion detection.

Commit:
Add the verbose argument to the function of checking qsub job completion.

Commit:
Remove the library call from the functions.

Commit:
Add qsub statements to the creation of the settings in case you are using qsub status detection functions.

Commit:
Relocate qsub functions for the analysis part and optimize the function design.

Commit:
Update the roxygen structure.

Commit:
Update dependency.

Commit:
Update and fix namespaces.

Commit:
Fix GitHub checks.

Commit:
Fix function usage.

Commit:
Replace read function.

infotroph commented on base/remote/R/check_qsub_status.R:
```diff
@@ -25,7 +26,9 @@ qsub_run_finished <- function(run, host, qstat) {
   if (length(out) > 0 && substring(out, nchar(out) - 3) == "DONE") {
-    PEcAn.logger::logger.debug("Job", run, "for run", run_id_string, "finished")
+    if (verbose) {
+      PEcAn.logger::logger.debug("Job", run, "for run", run_id_string, "finished")
```
PEcAn.logger already has built-in verbosity control via `logger.setLevel()` -- is a function-specific verbose flag needed here or is it enough for the user to set the logger level to something higher than debug so that this message isn't printed?

DongchenZ replied:
1) I am unsure if I fully understand how `logger.setLevel()` works inside the `PEcAn.logger` package.
2) For a small amount of jobs, I think it will still be helpful to have job info printed instead of creating a progress bar.

infotroph commented on modules/assim.sequential/R/Analysis_sda_block.R:
```diff
@@ -430,7 +439,7 @@ MCMC_block_function <- function(block) {
                   control = list(propCov= block$data$pf, adaptScaleOnly = TRUE,
                                  latents = "X", pfOptimizeNparticles = TRUE))
-
+  
```
```suggestion

```

infotroph committed:
Update modules/assim.sequential/R/Analysis_sda_block.R

infotroph commented on modules/assim.sequential/DESCRIPTION:
```diff
@@ -65,4 +67,4 @@ Suggests:
 Copyright: Authors
 Encoding: UTF-8
-RoxygenNote: 7.3.2
+RoxygenNote: 7.3.2
```
```suggestion
RoxygenNote: 7.3.2

```

infotroph committed:
Update modules/assim.sequential/DESCRIPTION

infotroph commented on modules/assim.sequential/R/Multi_Site_Constructors.R:
```diff
@@ -6,9 +6,6 @@
 ##' @param localization.FUN This is the function that performs the localization of the Pf matrix and it returns a localized matrix with the same dimensions.
-##' @param t not used
-##' @param blocked.dis passed to `localization.FUN`
-##' @param ... passed to `localization.FUN`
```
This is undoing a fix I made in #3346 (and apparently forgot to update Rcheck_reference.log, sorry! That's why the checks didn't complain about this being undone.)

infotroph replied:
...But if you have better descriptions for the parameters, please do improve my wording!

DongchenZ replied:
Fixed.

infotroph commented on modules/assim.sequential/R/Multi_Site_Constructors.R:
```diff
@@ -57,13 +54,13 @@ Contruct.Pf <- function(site.ids, var.names, X, localization.FUN=NULL, t=1, bloc
   
   return(pf.matrix.out)
-
+  
```
```suggestion

```

infotroph commented on modules/assim.sequential/R/Multi_Site_Constructors.R:
```diff
@@ -27,15 +24,15 @@ Contruct.Pf <- function(site.ids, var.names, X, localization.FUN=NULL, t=1, bloc
   site.cov.orders <- expand.grid(site.ids,site.ids) %>%
     dplyr::filter( .data$Var1 != .data$Var2)
-
+  
```
```suggestion

```

infotroph commented on modules/assim.sequential/R/Nimble_codes.R:
```diff
@@ -187,14 +187,14 @@ GEF.MultiSite.Nimble <-  nimbleCode({
       q[1:YN, 1:YN] ~ dwish(R = aq[1:YN, 1:YN], df = bq) ## aq and bq are estimated over time
     }
-
+    
```
```suggestion

```

infotroph commented on modules/assim.sequential/R/Nimble_codes.R:
```diff
@@ -187,14 +187,14 @@ GEF.MultiSite.Nimble <-  nimbleCode({
     ## add process error to x model but just for the state variables that we have data and H knows who
     X[1:YN]  ~ dmnorm(Xs[1:YN], prec = q[1:YN, 1:YN])
-
+    
```
```suggestion

```

infotroph commented on modules/assim.sequential/R/sda.batch.functions.R:
```diff
@@ -0,0 +1,212 @@
+  parallel::stopCluster(cl)
+  writeLines("finished",con = file.path(folder.path, "result.txt"))
+  save(results, file = file.path(folder.path, "results.Rdata"))
+}
```
```suggestion
}

```

infotroph commented on modules/assim.sequential/R/sda.enkf_NorthAmerica.R:
```diff
@@ -0,0 +1,381 @@
+#' @description State Variable Data Assimilation: Ensemble Kalman Filter and Generalized ensemble filter. Check out SDA_control function for more details on the control arguments.
+#' 
+#' @return NONE
+#' @import nimble
```
Can we importFrom just the functions we need instead of bringing all of nimble into the namespace?

DongchenZ replied:
I removed it cause I found there is no place where `nimble` is applied.

infotroph committed:
whitespace

Commit:
Update roxygen text.

Commit:
Revert "Update roxygen text."

This reverts commit 345530f48061c4ebfdaa909dd79720d62fd11cfb.

DongchenZ committed:
Merge branch 'PecanProject:develop' into develop

Commit:
Merge branch 'develop' of https://github.com/DongchenZ/pecan into develop

Commit:
Update roxygen.

Commit:
Revert the changes.

DongchenZ committed:
Merge branch 'PecanProject:develop' into develop

Commit:
Pass the printing when the job is finished.

Commit:
Bug fixes for the indexing when the scale is set to zero.

Commit:
Update the file deletion arguments.

Commit:
Add the feature of exporting all analysis and forecast results.

Commit:
Reduce the job time limit to request nodes more efficiently. Also, update the file deletion function even if there is no job.txt file.

Commit:
Remove the usage of the site name and the feature of parallel extraction due to the limitation of SoilGrids API.

Commit:
Attempt to fix github check issue.

Commit:
Attempt to fix github check issue.

Commit:
Resolve github check issue.

Commit:
Revert changes.

Commit:
Revert changes.

Commit:
Revert changes.

mdietze commented:
@DongchenZ says that the changes here were superseded by the now committed PR #3544 and that this PR can be closed
